alexa/ask-cli
# feat: rewrite of cf deployer

### Summary
full rewrite of CloudFormation Deployer
1) changed to async/await and promises
2) fixed error handling logic (fix issues when no error message is displayed when stack deploy fails)
3) removed is code modified check since it was not working. Correct checking if code is modified should be fixed in https://github.com/alexa/ask-cli/pull/245
4) added custom s3 bucket and key
5) added user provided CloudFormation parameters
6) added user provided CloudFormation capabilities
7) added CloudFormation outputs to .ask/ask-states.json file

**Automated Tests**
```
npm test
npm run functional-test
```

**Some Additional Manual Tests**
1) run deploy on existing skill - should succeed
```
==================== Deploy Skill Metadata ====================
[Warn]: The hash of current skill package folder does not change compared to the last deploy hash result, CLI will skip the deploy of skill package.
Skill ID: amzn1.ask.skill.c86c0da7-2350-48d5-981e-c9a0f1123f33

==================== Build Skill Code ====================
npm WARN hello-world@1.1.0 No repository field.

added 17 packages from 69 contributors and audited 17 packages in 3.182s
found 0 vulnerabilities

Skill code built successfully.
Code for region default built to /Users/kakuri/test/cf-skill2/.ask/lambda/build.zip successfully with build flow nodejs-npm.

==================== Deploy Skill Infrastructure ====================
  ✔ Deploy Alexa skill infrastructure for region "default"
Skill infrastructures deployed successfully through @ask-cli/cfn-deployer.

==================== Enable Skill ====================
Skill is already enabled, skip the enable process.
```
2) run deploy with incorrect setting in stack file - should fail with meaningful error message
```
==================== Deploy Skill Metadata ====================
...
==================== Build Skill Code ====================
...
==================== Deploy Skill Infrastructure ====================
  ✖ Deploy Alexa skill infrastructure for region "default"
[Error]: CliError: The CloudFormation deploy failed for Alexa region "default": AlexaSkillFunction[AWS::Lambda::Function]  UPDATE_FAILED (1 validation error detected: Value '512000' at 'memorySize' failed to satisfy constraint: Member must have value less than or equal to 3008 (Service: AWSLambdaInternal; Status Code: 400; Error Code: ValidationException; Request ID: e0c5c178-ff2b-4770-8619-ce22b11205fd; Proxy: null))
```
3)run deploy with wrong aws secret key in .aws/credentials - should fail with meaningful error message
```
==================== Deploy Skill Metadata ====================
...
==================== Build Skill Code ====================
...
==================== Deploy Skill Infrastructure ====================
  ✖ Deploy Alexa skill infrastructure for region "default"
[Error]: CliError: The CloudFormation deploy failed for Alexa region "default": The request signature we calculated does not match the signature you provided. Check your key and signing method.
```
4) run deploy with custom artifact s3, cfn parameters and capabilities
```
==================== Deploy Skill Metadata ====================
...
==================== Build Skill Code ====================
...
==================== Deploy Skill Infrastructure ====================
  ✔ Deploy Alexa skill infrastructure for region "default"
  The api endpoints of skill.json have been updated from the skill infrastructure deploy results.
Skill infrastructures deployed successfully through @ask-cli/cfn-deployer.

==================== Enable Skill ====================
Skill is already enabled, skip the enable process.
```
5)run deploy with s3 bucket in a different region - should fail with meaningful error message
```
==================== Deploy Skill Metadata ====================
...
==================== Build Skill Code ====================
...
==================== Deploy Skill Infrastructure ====================
  ✖ Deploy Alexa skill infrastructure for region "default"
[Error]: CliError: The CloudFormation deploy failed for Alexa region "default": AlexaSkillFunction[AWS::Lambda::Function]  UPDATE_FAILED (Error occurred while GetObject. S3 Error Code: PermanentRedirect. S3 Error Message: The bucket is in this region: us-west-1. Please use this region to retry the request (Service: AWSLambdaInternal; Status Code: 400; Error Code: InvalidParameterValueException; Request ID: 811b3344-2f8e-436d-9986-1fda88a4681d; Proxy: null))
```
